### PR TITLE
add multiprotocol handling to known_services

### PR DESCRIPTION
**Pull request by mauropalumbo75**
This pull request add a correct handling of multiple protocols and fix issue #419 .
Besides, it also remove logs with empty service from known_services.log (not sure about this, but it doesn't seem right to me to log a pair host/port if there was no service identified...).

Note that there is a slight difference between the output when using broker store or not (use_service_store= T or F). In the first case, use_service_store= T, the script logs multiple times the same host/port if the list of services is not exactly the same. For example, it can log:
10.0.0.1 445 HTTP
10.0.0.1 445 HTTP, SSL

In the second case, use_service_store= F, if a protocol has already been seen on a certain host/port, it won't be logged again. The above example will log as:
10.0.0.1 445 HTTP
10.0.0.1 445 SSL

i.e. the second line contains only SSL even if the originating conn has both HTTP and SSL.

### CODE REVIEW

File: scripts/policy/protocols/conn/known-services.zeek [hunk 1]
```diff
@@ -89,12 +111,16 @@ event zeek_init()
...
-	local key = AddrPortPair($host = info$host, $p = info$port_num);
+	local v : vector of string;
+	for ( s in info$service )
+		v += s;
+	sort(v, strcmp);	# sort the vector for proper key comparison in put_unique
+
+	local key = AddrPortServTriplet($host = info$host, $p = info$port_num, $serv = v);
 
 	when ( local r = Broker::put_unique(Known::service_store$store, key,
 	                                    T, Known::service_store_expiry) )
```
Comment (jsiwek) [inline scripts/policy/protocols/conn/known-services.zeek]: This data store seems only used for log de-duplication purposes at the moment, so you might just do a loop over each service and try `put_unique` for each key that is of the form `(host, port, single-service)`.  Changing it to that would then make the log output consistent with the other implementation of known-services that only uses partitioning/events -- e.g. both would log individual entries for either HTTP and SSL instead of having the data store version log just HTTP then also HTTP,SSL.

File: scripts/policy/protocols/conn/known-services.zeek [hunk 2]
```diff
@@ -210,10 +264,15 @@ event connection_state_remove(c: connection) &priority=-5
 	if ( c$known_services_done )
 		return;
 
-	if ( c$resp$state != TCP_ESTABLISHED )
+	# log tcp connection if established or udp if active
+	if ( c$resp$state != TCP_ESTABLISHED && c$resp$state != UDP_ACTIVE)
+		return;
```
Comment (jsiwek) [inline scripts/policy/protocols/conn/known-services.zeek]: Is this change directly related to multi-protocol service handling or something else?  May be better to separate into another PR for clarity if it's related to something else.

File: scripts/policy/protocols/conn/known-services.zeek [hunk 3]
```diff
@@ -58,15 +59,21 @@ export {
 	## :zeek:see:`Known::service_store`.
 	option service_store_timeout = 15sec;
 
+	## The timeout interval to wait after protocol confirmation before logging
+	## a new service. Some time is needed after a protocol confirmation event
+	## to either obtain a reply on certain protocols (i.e. DNS) or to fill
+	## :zeek:type:`connection$service`. 
+	option protocol_confirmation_timeout = 5min;
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/policy/protocols/conn/known-services.zeek]: No, this PR doesn't really depend on my other PR #490, sorry. I'll close this PR and reopen it without this change and also without the other changes below you want moved into a different PR.

File: scripts/policy/protocols/conn/known-services.zeek [hunk 4]
```diff
@@ -186,22 +233,29 @@ function known_services_done(c: connection)
...
+	# when checking for known_services on triplets (host, port, services). The service starting with "-" (i.e. -HTTP) is 
+	# reconized as different from the normal one (HTTP).
+	# It would be better to correct the analyzers some time later...
+	local tempservs : set[string];
+        for (s in c$service)
+        	if ( s[0] == "-" )
+			add tempservs[s[1:]];
+		else
+			add tempservs[s];
+
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/policy/protocols/conn/known-services.zeek]: ok, I didn't know that. So, if I understood correctly, if there is a "-HTTP" in c$service it means the HTTP protocol was first detected (protocol confirmation) and then the analyzer found out it was not it and rise a protocol violation event, right? If this is the case, the above code must be modified. I thought the "-" was just spurious.

File: scripts/policy/protocols/conn/known-services.zeek [hunk 5]
```diff
@@ -151,30 +194,34 @@ event Cluster::node_down(name: string, id: string)
...
 	if ( Known::use_service_store )
 		return;
 
-	if ( [info$host, info$port_num] in Known::services )
+	if ( check(info) )
 		return;
 
-	local key = cat(info$host, info$port_num);
+	local key = cat(info$host, info$port_num, info$service);
 	Cluster::publish_hrw(Cluster::proxy_pool, key, known_service_add, info);
```
[the remark this answers is not in the document]
Comment (mauropalumbo75) [reply scripts/policy/protocols/conn/known-services.zeek]: I haven't considered the case of a proxy pool with multiple proxies. Thanks for pointing this out. I'll change this in the new PR.

File: scripts/policy/protocols/conn/known-services.zeek (hunk 2 above)
Comment (mauropalumbo75) [reply scripts/policy/protocols/conn/known-services.zeek]: ok
